test/common/http/conn_pool_grid_test.cc

```diff
@@ -837,6 +845,69 @@ TEST_F(ConnectivityGridTest, SuccessWithoutHttp3NoMatchingAlpn) {
   grid_->callbacks()->onPoolReady(encoder_, host_, info_, absl::nullopt);
 }
 
+// Test the TCP pool will be immediately attempted if HTTP/3 has failed before.
+TEST_F(ConnectivityGridTest, Http3FailedRecentlyThenSucceeds) {
+  initialize();
+  addHttp3AlternateProtocol();
+  grid_->onZeroRttHandshakeFailed();
+  EXPECT_TRUE(ConnectivityGridForTest::hasHttp3FailedRecently(*grid_));
+  EXPECT_EQ(grid_->first(), nullptr);
+
+  // This timer will be returned and armed as the grid creates the wrapper's failover timer.
+  Event::MockTimer* failover_timer = new StrictMock<MockTimer>(&dispatcher_);
+  EXPECT_CALL(*failover_timer, enabled()).WillOnce(Return(false)).WillOnce(Return(true));
+  EXPECT_CALL(*failover_timer, enableTimer(_, _));
+  EXPECT_NE(grid_->newStream(decoder_, callbacks_,
+                             {/*can_send_early_data_=*/true,
+                              /*can_use_http3_=*/true}),
+            nullptr);
+  EXPECT_NE(grid_->first(), nullptr);
+  // The 2nd pool should be TCP pool and it should have been created together with h3 pool.
+  EXPECT_NE(grid_->second(), nullptr);
+  EXPECT_EQ(2u, grid_->callbacks_.size());
```

Comment:
I'm not quite sure I understand how this test that the TCP attempt is not delayed?

Reply:
Ah, thanks. I see.